### PR TITLE
Near complete refactor and rework of how forge crafting benches work

### PR DESCRIPTION
**Pull request by Paxilmaniac**
## About The Pull Request
Refactors forging workbenches to work off a list of datums given to it, rather than a hardcoded list of recipes and ingredients.

The bench now checks its immediate surroundings for ingredients, rather than using a forced set of ingredients within the hardcoded forge list, so (aside from weapons) you craft items by just having the ingredients next to the bench itself.

The menu was also improved, now being a radial menu. There are also much easier to read from descriptions for each recipe.

## How This Contributes To The Skyrat Roleplay Experience
Significantly improves the crafting bench's code, making it nicer and easier to use, and making it easier to use having never interacted with the forging system before with some improved descriptions.

## Proof of Testing
image
image
image

## Changelog

:cl:
qol: Crafting benches now take their resources from the tiles surrounding, rather than having to put them into the bench.
qol: The crafting bench UI is now a radial menu, and there are now several improved examines for see what ingredients or steps a recipe needs.
fix: All of the materials of the components for crafting things (forged components that is, not using wood) will be added to the weapon, rather than just the materials of the last item used, which means no more material deletion.
refactor: Crafting bench code has been changed to use datums rather than a hardcoded list of recipes
/:cl:

### CODE REVIEW

**Comment by Paxilmaniac** [comment]
the icebox commits are because I accidentally started this on a merge mirror branch locally, I made sure this didn't actually edit icebox in any way

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 1]
```diff
@@ -11,82 +9,85 @@
...
+	populate_radial_choice_list()
+
+/obj/structure/reagent_crafting_bench/proc/populate_radial_choice_list()
+	if(!length(allowed_choices))
+		return
+
+	if(length(radial_choice_list) && length(recipe_names_to_path)) // We already have both of these and don't need it, if this is called after these are generated for some reason
+		return
+
+	for(var/recipe as anything in allowed_choices)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]: unneeded as anything

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 2]
```diff
@@ -11,82 +9,85 @@
...
+		var/datum/crafting_bench_recipe/recipe_to_take_from = new recipe()
+		var/obj/recipe_resulting_item = recipe_to_take_from.resulting_item
+		radial_choice_list[recipe_to_take_from.recipe_name] = image(icon = initial(recipe_resulting_item.icon), icon_state = initial(recipe_resulting_item.icon_state))
+		recipe_names_to_path[recipe_to_take_from.recipe_name] = recipe
+		QDEL_NULL(recipe_to_take_from)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]: just needs to be qdel

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 3]
```diff
@@ -11,82 +9,85 @@
...
 
 /obj/structure/reagent_crafting_bench/examine(mob/user)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm:67]: add a `return .` at the bottom

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 4]
```diff
@@ -11,82 +9,85 @@
...
-	if(goal_name)
-		. += span_notice("Goal Item: [goal_name]")
-		. += span_notice("When you have the necessary materials, begin hammering!<br>")
+	if(!selected_recipe)
+		return
 
-		if(required_mats[CHAIN_RESOURCE])
-			. += span_warning("[required_mats[CHAIN_RESOURCE]] chains required.")
+	var/obj/resulting_item = selected_recipe.resulting_item
+	. += span_notice("Selected recipe's resulting item is: <b>[initial(resulting_item.name)]</b> <br>")
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]:
```suggestion
	. += span_notice("The selected recipe's resulting item is: <b>[initial(resulting_item.name)]</b> <br>")
```

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 5]
```diff
@@ -98,214 +99,244 @@
...
-			required_mats[WOOD_RESOURCE] = 4
-		if("Bow")
-			required_mats[WOOD_RESOURCE] = 4
-
-	if(!required_hits)
-		required_hits = (required_mats[CHAIN_RESOURCE] * 2) + (required_mats[PLATE_RESOURCE] * 2) + (required_mats[WOOD_RESOURCE] * 2)
-
-	balloon_alert(user, "choice made!")
-	update_appearance()
+	var/chosen_recipe = show_radial_menu(user, src, radial_choice_list, radius = 38, require_near = TRUE, tooltips = TRUE)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm:126]: oh my god this is so much better

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 6]
```diff
@@ -98,214 +99,244 @@
...
-		return FALSE
-
-	if(current_mats[PLATE_RESOURCE] < required_mats[PLATE_RESOURCE])
-		balloon_alert(user, "not enough materials!")
-		return FALSE
-
-	if(current_mats[WOOD_RESOURCE] < required_mats[WOOD_RESOURCE])
-		balloon_alert(user, "not enough materials!")
-		return FALSE
+	if(isnull(chosen_recipe))
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]:
```suggestion
	if(!chosen_recipe)
```

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 7]
```diff
@@ -98,214 +99,244 @@
...
-	for(var/i in 1 to current_mats[PLATE_RESOURCE])
-		new PLATE_RESOURCE(src_turf)
-
-	var/spawning_wood = current_mats[WOOD_RESOURCE] + 5
-	for(var/i in 1 to spawning_wood)
-		new WOOD_RESOURCE(src_turf)
-
-	qdel(src)
-	return TRUE
+	deconstruct(TRUE)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]: might want to use a named arg for this for readability

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 8]
```diff
@@ -98,214 +99,244 @@
...
-	var/skill_modifier = user.mind.get_skill_modifier(/datum/skill/smithing, SKILL_SPEED_MODIFIER) * 1 SECONDS
+	var/skill_modifier = user.mind.get_skill_modifier(selected_recipe.relevant_skill, SKILL_SPEED_MODIFIER) * 1 SECONDS
 
-	if(!COOLDOWN_FINISHED(src, hit_cooldown))
-		current_hits -= 3
-		balloon_alert(user, "bad hit!")
-		if(current_hits <= -required_hits)
-			clear_required()
+	if(!COOLDOWN_FINISHED(src, hit_cooldown)) // If you hit it before the cooldown is done, you get a bad hit, setting you back three good hits
+		current_hits_to_completion -= 3
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]: make a define

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 9]
```diff
@@ -98,214 +99,244 @@
...
-			spawned_obj.set_custom_materials(material_datum_ref)
 
-		balloon_alert(user, "item crafted!")
-		update_appearance()
-		user.mind.adjust_experience(/datum/skill/smithing, 30) //creating grants you something
-		current_mats[CHAIN_RESOURCE] -= required_mats[CHAIN_RESOURCE]
-		current_mats[PLATE_RESOURCE] -= required_mats[PLATE_RESOURCE]
-		current_mats[WOOD_RESOURCE] -= required_mats[WOOD_RESOURCE]
-		clear_required()
+	if(current_hits_to_completion >= selected_recipe.required_good_hits && !length(contents))
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]:
```suggestion
	if((current_hits_to_completion >= selected_recipe.required_good_hits) && !length(contents))
```
clarity and i don't trust DM not to fuck this up in the future

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 10]
```diff
@@ -98,214 +99,244 @@
...
+
+			if(isstack(nearby_item)) // If the item is a stack, check if that stack has enough material in it to fill out the amount
+				var/obj/item/stack/nearby_stack = nearby_item
+				required_amount -= nearby_stack.amount
+			else // Otherwise, we still exist and should subtract one from the required number of items
+				required_amount -= 1
+
+			requirement_items += nearby_item
+
+		if(required_amount > 0)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm:252]:
```suggestion
		if(required_amount)
```

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 11]
```diff
@@ -98,214 +99,244 @@
...
+		recipe_to_follow = new /datum/crafting_bench_recipe/weapon_completion_recipe
+	var/list/materials_to_transfer = use_or_delete_recipe_requirements(things_to_use, recipe_to_follow)
+
+	var/obj/newly_created_thing
+
+	if(completing_a_weapon)
+		var/obj/item/forging/complete/completed_forge_item = contents[1]
+		newly_created_thing = new completed_forge_item.spawning_item(src)
+		if(completed_forge_item.custom_materials) // We need to add the weapon head's materials to the completed item, too
+			for(var/custom_material as anything in completed_forge_item.custom_materials)
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm]: unneded as anything

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm [hunk 12]
```diff
@@ -98,214 +99,244 @@
...
+/obj/structure/reagent_crafting_bench/proc/get_environment()
+	. = list()
+
+	if(!get_turf(src))
+		return
+
+	for(var/atom/movable/found_movable_atom in range(1, src))
+		if((found_movable_atom.flags_1 & HOLOGRAM_1))
+			continue
+		. += found_movable_atom
```
Comment (Zonespace27) [inline modular_skyrat/modules/reagent_forging/code/crafting_bench.dm:353]: return .

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
Late night review as promised

File: modular_skyrat/modules/reagent_forging/code/crafting_bench.dm (hunk 3 above)
Comment (Paxilmaniac) [reply modular_skyrat/modules/reagent_forging/code/crafting_bench.dm:67]: I don't see the use in that?

Comment (Paxilmaniac) [reply modular_skyrat/modules/reagent_forging/code/crafting_bench.dm:67]: Oh you mean for the examine proc don't you, I don't see literally any other examine procs do that so idk about that one.

**Review by Zonespace27 — APPROVED** [review]

**Comment by GoldenAlpharex** [comment]
Gonna blame this on Zonespace if anything wrong happens :)

**Pull request merged**